// modules/feedback_list/API/loadFeedbacks.js
import { API_URL } from 'constants'

export default async function loadFeedbacks({personId, feedbackType, limit}) {
    const response = await fetch(`${API_URL}/feedback`)
    const jsonResponse = await response.json()

    return jsonResponse.map(feedback => {
        return {
            id: feedback.id,
            user: {
                userName: feedback.user.userName,
                createDate: feedback.createDate,
                userImage: feedback.userImage
            },
            rating: feedback.rating,
            feedbackBody: feedback.body,
            media: [...feedback.media],
            comments: feedback.comments.map(comment => {
                return {
                    id: comment.id,
                    user: {
                        userName: comment.user.userName,
                        createDate: comment.user.createDate,
                        userImage: comment.user.userImage
                    },
                    commentBody: comment.commentBody
                }
        })
        }
    })
}
